### PR TITLE
Updated Independence Tests - Testing

### PR DESCRIPTION
**Pull request by ewengillies**
This merge request is the first step in updating the `indepdence_tests.py` module.  To do so, this request covers the following changes:
 * Change tests to `pytest` and update
 * Fix any bugs found from these tests
 * Plan changes to `indepdence_tests.py` code.

## Change tests to `pytest` and update

Previous to this merge request, the `unittest` module was used for all tests.  As with the other tests in this project, the existing tests have been moved to `pytest`.  Mainly using pytests fixtures and parameters, several key features are now much more thoroughly tested.

### Testing `_construct_array`

Previous to this request, the tests for `_construct_array` did not cover many use cases for this core function.  The cases covered also did not seem to test thoroughly either.  To improve this, the possible use cases are covered using several functions/fixtures.

 * [rand_node()](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L30) : Generate a random node within given `node_id` and `tau` limits.
 *  [gen_nodes()](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L38) : Generate a collection of nodes
    * Returns `x_nds, y_nds, z_nds`.
    * `y_nds` is always just `[(0, 0)]`
    * `x_nds` is always just one random node
    * `z_nds` is a variable number of random nodes.
  * [cstrct_array_params()](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L81) : `pytest` fixture to handle parameters for `_construct_array`
    * Parameterized to cover all use cases.
    * Many different node combinations tested. 
    * Missing values tested, with and without masking.
    * All mask types tested independently, and all types together.

These are used to test `_construct_array()` in [test_construct_array](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests_/test_independence_tests.py#L84).  This test has several stages to ensure proper results can be compared to the generated arrays:
  * Fake data is generated, with length 15 in time for 10 nodes. 
  * If a mask is used, then the first value in time on the first node of that type is masked.
  * If missing values are used, then a missing value flag is inserted into the first non-masked z-node.  
  * To compare results, `x_nds`, `y_nds`, and `z_nds` are all made unique.  
  * The expected values of the array are generated using knowledge of how the original array was constructed.
  * The resulting `array` and `xyz` are compared to the expected values.

These test all pass.

### Testing the independence tests 

With `_construct_array` tested, the independence tests themselves were tested.  This follows mainly from the existing testing.  A data generation function was defined for all independence tests, and a data sample and independence test fixture was defined for each independence test:
  * Data generation: 
    * [gen_data_sample()](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L52) : Generates 2D multivariate normal random data from a seed and off-diagonal correlation matrix element strength.
    * [data_sample_a](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L176) : Data fixture for sample for `ParCorr`
    * [data_sample_b](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L272) : Data fixture for sample for `GPDC`
    * [data_sample_c](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L355) : Data fixture for sample for `CMIknn`
    * [data_sample_d](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L402) : Data fixture for sample for `CMIsymb`
  * Independence test fixtures (currently unparameterized). 
    * [par_corr](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L155) : Fixture to build `ParCorr` independence test.  
    * [gdpc](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L253) : Fixture to build `GDPC` independence test.  
    * [cmi_knn](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L336) : Fixture to build `CMIknn` independence test.  
    * [cmi_symb](https://github.com/ewengillies/tigramite/blob/f59c8a7082995ce4bd29920c4417c95beea55b26/tests/test_independence_tests.py#L384) : Fixture to build `CMIsymb` independence test.  

Using these fixture, the following tests were run, albeit not over all independence tests:
  * Ensuring bootstrap and analytic confidence limits are the same : `ParCorr`
  * Ensuring shuffle significance and analytic significance are the same : `ParCorr`, `GPDC`
  * Checking the residuals are correct : `ParCorr`, `GPDC`
  * Checking the dependence measure is correct : `ParCorr`, `CMIknn`, `CMIsymb`

These tests all pass, but the test cases should be expanded to cover all independence tests.

## Fixing bugs uncovered by tests

The main "bug" found was in missing value flagging.  This function was fixed.  This function now uses `np.roll` to collect all the time slices that come withing `max_lag` of a missing value.  These time slices are removed.  

## Plans to change `independence_test.py`

Many `TODO`s were inserted to this effect.  These aim at two main motivations: 
 * Do not use member values to cache algorithm settings or algorithm results.
 * Use abstract base classes more effectively, inheriting from `ABCmeta` to increase their abilities.

### CODE REVIEW

File: tigramite/independence_tests.py [hunk 1]
```diff
@@ -129,106 +132,127 @@ def _construct_array(X, Y, Z, tau_max, data,
...
+        missing_anywhere = np.any(array == missing_flag, axis=0)
+        # Add on some dummy indices so we can permute the values across all
+        # allowed lags using np.roll without np.roll causing late-values to
+        # interfere with early values
+        missing_anywhere = np.append(missing_anywhere,
+                                     np.zeros((max_lag), dtype=bool))
+        # TODO check with jakob:
+        #   * Before, only lags up to tau were included.
         for tau in range(max_lag+1):
-            use_indices[missing_anywhere[tau:T-max_lag+tau]] = 0
```
Comment (ewengillies) [inline tigramite/independence_tests.py]: Change this code back to using data then transforming to array as in example

File: tigramite/independence_tests.py [hunk 2]
```diff
@@ -305,76 +329,87 @@ class CondIndTest(object):
...
         self.recycle_residuals = recycle_residuals
         if self.recycle_residuals:
             self.residuals = {}
-
+        # If we use a mask, do not recycle residuals
+        # TODO ask jakob:
+        #  * why does this come after the recycle residuals option is already
+        #  used?
+        #  * why does mask usage preclude the recycling of residuals?
         if self.use_mask:
```
Comment (ewengillies) [inline tigramite/independence_tests.py:349]: Mask Type == None can now take over for use_mask.  Move mask_type over to data frame.

File: tigramite/independence_tests.py [hunk 3]
```diff
@@ -305,76 +329,87 @@ class CondIndTest(object):
...
             self.recycle_residuals = False
```
Comment (ewengillies) [inline tigramite/independence_tests.py:350]: warn users that recycle residuals turns off if a mask is used.

File: tigramite/independence_tests.py (hunk 1 above)
Comment (ewengillies) [reply tigramite/independence_tests.py]: fixed the code, updated the tests

File: tigramite/independence_tests.py (hunk 2 above)
Comment (ewengillies) [reply tigramite/independence_tests.py:349]: fixing this in #16

File: tigramite/independence_tests.py (hunk 3 above)
Comment (ewengillies) [reply tigramite/independence_tests.py:350]: fixed in #16

**Review by ewengillies — COMMENTED** [review]
* fixed all discussed issues.  
 * tests run fine on local machine.  
 * testing on remote CI site (travis) is down at the moment.  
  * will be ready to merge once site is back up and working @jakobrunge

**Comment by ewengillies** [comment]
all tests passing on trello, ready to be merged @jakobrunge

**Review by jakobrunge — APPROVED** [review]

**Pull request merged**